transports/tcp/src/lib.rs

```diff
@@ -123,60 +128,77 @@ impl TcpConfig {
 impl Transport for TcpConfig {
     type Output = TcpTransStream;
     type Error = io::Error;
-    type Listener = TcpListenStream;
-    type ListenerUpgrade = FutureResult<Self::Output, io::Error>;
+    type Listener = TcpListener;
+    type ListenerUpgrade = FutureResult<Self::Output, Self::Error>;
     type Dial = TcpDialFut;
 
     fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
-        if let Ok(socket_addr) = multiaddr_to_socketaddr(&addr) {
-            let listener = TcpListener::bind(&socket_addr);
-            // We need to build the `Multiaddr` to return from this function. If an error happened,
-            // just return the original multiaddr.
-            let listen_addr = match listener {
-                Ok(ref l) => if let Ok(new_s_addr) = l.local_addr() {
-                    new_s_addr.to_multiaddr()
-                        .expect("multiaddr generated from socket addr is always valid")
-                } else {
-                    addr
-                },
-                Err(_) => addr,
+        let socket_addr =
+            if let Ok(sa) = multiaddr_to_socketaddr(&addr) {
+                sa
+            } else {
+                return Err(TransportError::MultiaddrNotSupported(addr))
             };
 
-            debug!("Now listening on {}", listen_addr);
-            let sleep_on_error = self.sleep_on_error;
-            let inner = listener
-                .map_err(TransportError::Other)?
-                .incoming()
-                .sleep_on_error(sleep_on_error);
-
-            Ok(TcpListenStream {
-                inner: Ok(inner),
-                listen_addr,
-                config: self,
-                tell_new_addr: true
-            })
-        } else {
-            Err(TransportError::MultiaddrNotSupported(addr))
-        }
+        let listener = tokio_tcp::TcpListener::bind(&socket_addr).map_err(TransportError::Other)?;
+        let local_addr = listener.local_addr().map_err(TransportError::Other)?;
+        let port = local_addr.port();
+
+        // Determine all our listen addresses which is either a single local IP address
+        // or (if a wildcard IP address was used) the addresses of all our interfaces,
+        // as reported by `get_if_addrs`.
+        let (addrs, new_addrs) =
+            if socket_addr.ip().is_unspecified() {
+                let addrs = host_addresses(port).map_err(TransportError::Other)?;
+                debug!("Listening on {:?}", addrs.values());
+                let new_addrs = addrs.values()
+                    .cloned()
+                    .map(ListenerEvent::NewAddress)
+                    .collect::<Vec<_>>();
+                (Either::Right(addrs), Either::Right(stream::iter_ok(new_addrs)))
+            } else {
+                let ma = local_addr.to_multiaddr().expect("A socket address is a valid Multiaddr.");
```

Comment:
I think you need to append the port here.

Reply:
`local_addr` is a `SocketAddr` which includes the port number, which is either the one provided in the original multi-address or, if a wildcard port number was used, the one that `bind` selected.